pyensembl/serializable.py

```diff
@@ -0,0 +1,143 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+
+class Serializable(object):
+    """
+    Base class for all PyEnsembl objects which provides default
+    methods such as to_json, from_json, __reduce__, and from_dict
+    and relies only on (1) a user-defined to_dict method and (2) that the
+    keys of to_dict match the arguments to __init__.
+    """
+
+    def __str__(self):
+        fields_str = ", ".join(
+            "%s=%s" % (k, v) for (k, v) in sorted(self.to_dict().items()))
+        return "%s(%s)" % (self.__class__.__name__, fields_str)
+
+    def __repr__(self):
+        return str(self)
+
+    def __eq__(self, other):
+        if self.__class__ is not other.__class__:
+            return False
+        self_dict = self.to_dict()
+        other_dict = other.to_dict()
+        if len(self_dict) != other_dict(other_dict):
+            return False
+        self_keys = set(self_dict.keys())
+        other_keys = set(other_dict.keys())
+        if self_keys != other_keys:
+            return False
+        for key in self_keys:
+            if self_dict[key] != other_dict[key]:
+                return False
+        return True
+
+    def to_dict(self):
+        raise NotImplementedError("Method to_dict() not implemented for %s" % (
+            self.__class__.__name__,))
+
+    @classmethod
+    def _tuple_to_class(cls, module_and_class_name):
+        """
+        Given the name of a module and a class it contains, imports that module
+        and gets the class object from it.
+        """
+        module_string, class_name = module_and_class_name
+        module = __import__(module_string)
+        return getattr(module, class_name)
+
+    @classmethod
+    def _class_to_tuple(cls):
+        """
+        Given an class, return two strings:
+            - fully qualified import path for its module
+            - name of the class
+
+        The class can be reconstructed from these two strings by calling
+        _tuple_to_class.
+        """
+        return cls.__module__, cls.__name__
+
+    def _object_to_primitive_types(self):
+        """
+        Given an instance of a Python object, returns a tuple whose
+        first element is a primitive representation of the class and whose
+        second element is a dictionary of instance data.
+        """
+        class_representation = self._class_to_tuple()
+        state_dict = self.to_dict()
+        return (class_representation, state_dict)
+
+    @classmethod
+    def _object_from_primitive_types(cls, class_and_data_pair):
```

Comment:
Nit: maybe `_primitive_types_to_object` to be consistent with e.g. `_class_to_tuple` vs. `_tuple_to_class`?